plutus-tx-plugin/executables/profile/Main.hs

```diff
@@ -24,12 +42,120 @@ import qualified PlutusCore.Default        as PLC
+-- the plutus-tx-plugin/executables/profile/ directory.
+writeLogToFile ::
+  ToUPlc a PLC.DefaultUni PLC.DefaultFun =>
+  -- | Name of the file you want to save it as.
+  FilePath ->
+  -- | The program to be profiled.
+  [a] ->
+  IO ()
+writeLogToFile fileName values = do
+  let filePath = "plutus-tx-plugin/executables/profile/"<>fileName<>".timelog"
+  log <- T.intercalate "\n" . view _2 <$> (rethrow $ runUPlcProfileExec values)
+  writeFile filePath (T.unpack log)
+  processed <- processLog filePath
+  writeFile (filePath<>".stacks") (printProcessedLog processed)
+  pure ()
+
+data ProfileEvent =
+  MkProfileEvent UTCTime Transition T.Text
+
+data Transition =
+  Enter
+  | Exit
+
+data StackTime =
+  MkStackTime [T.Text] Double
+
+processLog :: FilePath -> IO [StackTime]
+processLog file = do
+  content <- readFile file
+  let lEvents = lines content -- turn to a list of events
+  pure $ getStacks (profileEvents lEvents)
+
+printProcessedLog :: [StackTime] -> String
+printProcessedLog ((MkStackTime fns duration):tl) =
+    intercalate
+      "; "
+      -- reverse to make the functions in the order correct for flamegraphs.
+      (reverse (map T.unpack fns))
+      <>" "
+      -- multiplying the duration by a large number
+      -- to make the number show in a way flamegraph accepts.
+      <>show (duration*1000000)
+      <>"\n"<>printProcessedLog tl
+printProcessedLog _ = ""
+
+parseProfileEvent :: String -> ProfileEvent
+parseProfileEvent str =
+  case words str of
+    [t1,t2,t3,transition,var] ->
+      case transition of
+        "entering" -> MkProfileEvent (read (unwords [t1,t2,t3])::UTCTime) Enter (T.pack var)
+        "exiting" -> MkProfileEvent (read (unwords [t1,t2,t3])::UTCTime) Exit (T.pack var)
+        badLog -> error $
+          "parseProfileEvent: expecting \"entering\" or \"exiting\" but got "
+          <> show badLog
+    invalid -> error $
+      "parseProfileEvent: invalid log, expecting a form of [t1,t2,t3,transition,var] but got "
+      <> show invalid
+
+profileEvents :: [String] -> [ProfileEvent]
+profileEvents = map parseProfileEvent
+
+getStacks :: [ProfileEvent] -> [StackTime]
+getStacks = go []
+  where
+    go ::
+      [StackFrame] ->
+      [ProfileEvent] ->
+      [StackTime]
+    go curStack ((MkProfileEvent startTime Enter varName):tl) =
+          go
+            (MkStackFrame{varName, startTime, timeSpentCalledFn = 0}:curStack)
+            tl
+    go (MkStackFrame {varName=curTopVar, startTime, timeSpentCalledFn}:poppedStack) ((MkProfileEvent exitTime Exit var):tl)
+      | curTopVar == var =
+        let duration = diffUTCTime exitTime startTime
+            updateTimeSpent (hd@MkStackFrame{timeSpentCalledFn}:tl) =
+              hd {timeSpentCalledFn = timeSpentCalledFn + duration}:tl
+            updateTimeSpent [] = []
+            updatedStack = updateTimeSpent poppedStack
+            fnsEntered = map varName updatedStack -- this is quadratic but it's fine for fg
```

Comment:
Why is it fine? I think the answer is "because we have to do quadratic work anyway and the input sizes are small"